### PR TITLE
CylindricalBinaryCompactObject domain

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

Adds a DomainCreator called CylindricalBinaryCompactObject, which creates a binary domain based on cubed cylinders.

Also includes other smaller commits necessary for the new DomainCreator to work, or for making debugging it easier. These smaller commits are:
 - Improve error messages in Block.cpp and CylindricalEndcap.
 - Add CylindricalEndcap and other maps to RegisterDerivedWithCharm.cpp
 - Relax some parameter restrictions in CylindricalEndcap (these were tested with --repeat-until-fail 10000)
 - Add new helper functions to DomainHelpers for constructing cylindrical domains.

NOTE: CylindricalBinaryCompactObject is appropriate only for domains
that represent roughly equal-sized objects. Support for sufficiently
unequal-sized objects requires another map that doesn't exist yet.
CylindricalBinaryCompactObject currently errors when used for sufficiently
unequal-sized objects. Code that adds support for unequal-sized objects will
be added in a forthcoming PR.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by markscheel** [comment]
The failing test is an unrelated Unit.Time.Stepchoosers.ErrorControl failure that apparently chose an inconvenient random number.

File: src/Domain/DomainHelpers.hpp
Comment (wthrowe) [inline src/Domain/DomainHelpers.hpp]: Use math mode or backtick mode for the code stuff in this documentation.

File: src/Domain/DomainHelpers.hpp
Comment (wthrowe) [inline src/Domain/DomainHelpers.hpp]: Remove all the top-level consts in the non-definitions.

File: src/Domain/DomainHelpers.hpp
Comment (wthrowe) [inline src/Domain/DomainHelpers.hpp]: Could any of the code for these functions be shared with cyl_wedge_coordinate_maps?  I'm going to hold off on reviewing the implementations until we've decided this.

File: tests/Unit/Helpers/Domain/DomainTestHelpers.hpp
Comment (wthrowe) [inline tests/Unit/Helpers/Domain/DomainTestHelpers.hpp]: It takes a vector of blocks.

File: tests/Unit/Helpers/Domain/DomainTestHelpers.cpp
Comment (wthrowe) [inline tests/Unit/Helpers/Domain/DomainTestHelpers.cpp:482]: Keep the order of the functions in the hpp and cpp the same.

File: tests/Unit/Helpers/Domain/DomainTestHelpers.cpp
Comment (wthrowe) [inline tests/Unit/Helpers/Domain/DomainTestHelpers.cpp]: `index`

File: tests/Unit/Helpers/Domain/DomainTestHelpers.cpp
Comment (wthrowe) [inline tests/Unit/Helpers/Domain/DomainTestHelpers.cpp]: Could this be done more simply with VolumeCornerIterator?

File: tests/Unit/Helpers/Domain/DomainTestHelpers.cpp
Comment (wthrowe) [inline tests/Unit/Helpers/Domain/DomainTestHelpers.cpp]: Shouldn't need the `<>`.

File: src/Domain/Creators/CMakeLists.txt
Comment (wthrowe) [inline src/Domain/Creators/CMakeLists.txt]: Alphabetize.

File: src/Domain/Creators/CylindricalBinaryCompactObject.hpp
Comment (wthrowe) [inline src/Domain/Creators/CylindricalBinaryCompactObject.hpp]: `\le` and similar below.

File: src/Domain/Creators/CylindricalBinaryCompactObject.hpp
Comment (wthrowe) [inline src/Domain/Creators/CylindricalBinaryCompactObject.hpp]: I don't understand the connection between the equation and the text.

File: src/Domain/Creators/CylindricalBinaryCompactObject.cpp
Comment (wthrowe) [inline src/Domain/Creators/CylindricalBinaryCompactObject.cpp]: `const` argument.

File: src/Domain/Creators/CylindricalBinaryCompactObject.cpp
Comment (wthrowe) [inline src/Domain/Creators/CylindricalBinaryCompactObject.cpp]: `const` argument.

File: src/Domain/Creators/CylindricalBinaryCompactObject.cpp
Comment (wthrowe) [inline src/Domain/Creators/CylindricalBinaryCompactObject.cpp]: Use the input-file-names for user error messages.  ("CenterA" here, more below.)

File: src/Domain/Creators/CylindricalBinaryCompactObject.cpp
Comment (wthrowe) [inline src/Domain/Creators/CylindricalBinaryCompactObject.cpp]: It looks like you could just initialize these to the correct values without the lambda.

File: tests/Unit/Domain/Creators/Test_CylindricalBinaryCompactObject.cpp
Comment (wthrowe) [inline tests/Unit/Domain/Creators/Test_CylindricalBinaryCompactObject.cpp:318]: Note: If #3084 is merged before this, these two lines should become
```c++
      return TestHelpers::test_option_tag<
￼          domain::OptionTags::DomainCreator<3>,
```

File: tests/Unit/Domain/Creators/Test_CylindricalBinaryCompactObject.cpp
Comment (wthrowe) [inline tests/Unit/Domain/Creators/Test_CylindricalBinaryCompactObject.cpp]: Not particularly related to factory.

**Review by wthrowe — CHANGES_REQUESTED** [review]
I have not gone through and compared everything to the math in the SpEC paper.  If people think that is a valuable endeavor I can do it later.

File: src/Domain/Creators/CylindricalBinaryCompactObject.cpp
Comment (markscheel) [reply src/Domain/Creators/CylindricalBinaryCompactObject.cpp]: Yeah, the lambda was there from the version of this code that also handled higher mass ratios (and the interior of the lambda had some 'if's that handled both cases, so the lambda was needed.)  Removing the lambda.

File: tests/Unit/Helpers/Domain/DomainTestHelpers.cpp
Comment (markscheel) [reply tests/Unit/Helpers/Domain/DomainTestHelpers.cpp]: The loops would be simpler (one less loop), but VolumeCornerIterator returns a `std::array<double,VolumeDim>` instead of a `tnsr::I<double,VolumeDim,Frame::Logical>`, so using VolumeCornerIterator would mean that I need to convert the array to a tensor.

So for now I haven't changed this. Please let me know if I should.

File: src/Domain/DomainHelpers.hpp
Comment (markscheel) [reply src/Domain/DomainHelpers.hpp]: Ok, I did this as a separate fixup commit.

**Comment by markscheel** [comment]
Made changes (except VolumeCornerIterator).

File: tests/Unit/Helpers/Domain/DomainTestHelpers.cpp
Comment (nilsvu) [reply tests/Unit/Helpers/Domain/DomainTestHelpers.cpp]: FYI, there's also `Domain/Structure/Hypercube.hpp`. I think `VolumeCornerIterator` and `FaceCornerIterator` could probably be expressed as a `HypercubeElementsIterator`. Just letting you know it exists :)

File: src/Domain/DomainHelpers.hpp
Comment (wthrowe) [inline src/Domain/DomainHelpers.hpp]: How about just using a bool?

File: tests/Unit/Helpers/Domain/DomainTestHelpers.cpp
Comment (wthrowe) [reply tests/Unit/Helpers/Domain/DomainTestHelpers.cpp]: I think the conversion can be done inline with a cast or a call to a Tensor constructor.  If not, as-is is fine.  No preference on my part between VCI and Hypercube.

**Comment by markscheel** [comment]
More changes, rebased onto develop.

**Comment by wthrowe** [comment]
Looks good.  Squash.

**Comment by markscheel** [comment]
Squashed

**Review by wthrowe — APPROVED** [review]

**Comment by wthrowe** [comment]
(Reminder to myself not to merge this because it still needs a second review.)

File: src/Domain/Creators/CylindricalBinaryCompactObject.hpp
Comment (nilsdeppe) [inline src/Domain/Creators/CylindricalBinaryCompactObject.hpp:139]: Are the centers assumed to be on the x-axis? Do we want to impose that restriction?

File: src/Domain/Creators/CylindricalBinaryCompactObject.hpp
Comment (nilsdeppe) [inline src/Domain/Creators/CylindricalBinaryCompactObject.hpp:205]: The other binary domain changed the way it handled the two objects by having an `Object` class for each that holds the options. I'm not requesting that you change to that, just letting you know so you can decide what you prefer :) Not changing, changing in a later PR, or changing in this PR are all equally reasonable to me :)

**Review by nilsdeppe — APPROVED** [review]
LGTM, a couple comments I'd like your thoughts on but nothing that should hold this PR up :)

File: src/Domain/Creators/CylindricalBinaryCompactObject.hpp
Comment (markscheel) [reply src/Domain/Creators/CylindricalBinaryCompactObject.hpp:139]: The centers can and will be displaced in the y-direction, as is necessary for the initial data so that the initial ADM linear momentum is zero.  The Domain also accommodates offsets in the z-direction, so we don't need to impose that restriction either (but I think in practice all our runs have z=0 for both objects, so far).

File: src/Domain/Creators/CylindricalBinaryCompactObject.hpp
Comment (markscheel) [reply src/Domain/Creators/CylindricalBinaryCompactObject.hpp:205]: Here there are only two options (Center and Radius) that are different for the two objects.  Perhaps such a change can be made if/when there are more such options?  One thing that will probably be needed in a future PR is different refinement levels in different Blocks (the other binary domain has this feature).

File: src/Domain/Creators/CylindricalBinaryCompactObject.hpp
Comment (nilsdeppe) [reply src/Domain/Creators/CylindricalBinaryCompactObject.hpp:139]: ah okay, thanks!

File: src/Domain/Creators/CylindricalBinaryCompactObject.hpp
Comment (nilsdeppe) [reply src/Domain/Creators/CylindricalBinaryCompactObject.hpp:205]: Yep, sounds good :)

**Pull request merged**
